File: sentimental-credit/credit.py
def main():
    while True:
        number = input("Number: ")

        if not number.isdigit():
            continue
        break

    number = int(number)

    if number < 0:
        print("INVALID")
        return

    isValid = CardNumberValidation(number)

    if isValid and MasterCardValidation(number):
        print("MASTERCARD")
        return
    elif isValid and AmericanExpressCardValidation(number):
        print("AMEX")
        return
    elif isValid and VisaCardValidation(number):
        print("VISA")
        return
    else:
        print("INVALID")
        return


# Function for validate card numbers
def CardNumberValidation(cardNumber: int) -> bool:
    decimal = 0
    i = 1
    secondResult = 0
    while decimal < cardNumber:
        fix = 10 * i
        decimal = 100 * i
        secondDigit = cardNumber % decimal // fix * 2
        i *= 100

        secondResult += secondDigit // 10 + secondDigit % 10

    i = 1
    decimal = 0
    sum = secondResult
    while decimal < cardNumber:
        fix = 1 * i
        decimal = 10 * i
        digit = cardNumber % decimal // fix
        i *= 100

        sum += digit

    if sum % 10 == 0:
        return True
    return False

# Differents companies cards validation functions
# Start


def MasterCardValidation(card: int) -> bool:
    CSNu1 = 51
    CSNu2 = 52
    CSNu3 = 53
    CSNu4 = 54
    CSNu5 = 55

    Sn = card // pow(10, 14)

    return len(str(card)) == 16 and (
        Sn == CSNu1 or Sn == CSNu2 or
        Sn == CSNu3 or Sn == CSNu4 or
        Sn == CSNu5
    )


def AmericanExpressCardValidation(card: int) -> bool:
    CSNu1 = 34
    CSNu2 = 37

    Sn = card // pow(10, 13)

    return len(str(card)) == 15 and (
        Sn == CSNu1 or Sn == CSNu2
    )


def VisaCardValidation(card: int) -> bool:
    CSNu1 = 4
    is16H = len(str(card)) == 16
    is13H = len(str(card)) == 13
    Sn = int()

    if is16H:
        Sn = card // pow(10, 15)
    elif is13H:
        Sn = card // pow(10, 12)

    return (is16H or is13H) and Sn == CSNu1
# End


main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool CardNumberValidation(long cardNumber);

long power(int base, int power);

bool NumbersQuantityCheck(long card, int n);

bool MasterCardValidation(long card);

bool AmericanExpressCardValidation(long card);

bool VisaCardValidation(long card);

int main(void)
{
    long card = get_long("Number: ");
    if (card < 0)
    {
        printf("INVALID");
        return 0;
    }

    bool valid = CardNumberValidation(card);
    if (valid && MasterCardValidation(card))
    {
        printf("MASTERCARD\n");
        return 0;
    }
    else if (valid && AmericanExpressCardValidation(card))
    {
        printf("AMEX\n");
        return 0;
    }
    else if (valid && VisaCardValidation(card))
    {
        printf("VISA\n");
        return 0;
    }
    else
    {
        printf("INVALID\n");
        return 0;
    }
}

long power(int base, int power)
{
    long result = base;
    for (int i = 1; i < power; i++)
    {
        result = result * base;
    }
    return result;
}

bool MasterCardValidation(long card)
{
    const int CorrectStartNumber1 = 51;
    const int CorrectStartNumber2 = 52;
    const int CorrectStartNumber3 = 53;
    const int CorrectStartNumber4 = 54;
    const int CorrectStartNumber5 = 55;

    int startNumber = card / power(10, 14);

    return NumbersQuantityCheck(card, 16) &&
           (startNumber == CorrectStartNumber1 || startNumber == CorrectStartNumber2 ||
            startNumber == CorrectStartNumber3 || startNumber == CorrectStartNumber4 ||
            startNumber == CorrectStartNumber5);
}

bool AmericanExpressCardValidation(long card)
{
    const int CorrectStartNumber1 = 34;
    const int CorrectStartNumber2 = 37;

    int startNumber = card / power(10, 13);

    return NumbersQuantityCheck(card, 15) &&
           (startNumber == CorrectStartNumber1 || startNumber == CorrectStartNumber2);
}

bool VisaCardValidation(long card)
{
    const int CorrectStartNumber1 = 4;
    int startNumber = 0;
    bool numbers16Check = NumbersQuantityCheck(card, 16);
    bool numbers13Check = NumbersQuantityCheck(card, 13);

    if (numbers16Check)
    {
        startNumber = card / power(10, 15);
    }
    else if (numbers13Check)
    {
        startNumber = card / power(10, 12);
    }

    return (numbers13Check || numbers16Check) && startNumber == CorrectStartNumber1;
}

bool NumbersQuantityCheck(long card, int n)
{
    int decimalHouses = 0;
    long number = card;
    for (int i = 1; i <= n; i++)
    {
        decimalHouses = i;
        number = number / 10;
        if (number == 0)
        {
            break;
        }
    }
    return n == decimalHouses;
}

bool CardNumberValidation(long cardNumber)
{
    long decimal = 0;
    long fix = 0;
    long i = 1;
    int secondResult = 0;
    int secondDigit;
    int digit;
    while (decimal < cardNumber)
    {
        fix = 10 * i;
        decimal = 100 * i;
        secondDigit = cardNumber % decimal / fix * 2;

        secondResult += secondDigit / 10 + secondDigit % 10;
        i *= 100;
    }

    i = 1;
    decimal = 0;
    int sum = secondResult;
    while (decimal < cardNumber)
    {
        fix = 1 * i;
        decimal = 10 * i;
        digit = cardNumber % decimal / fix;

        sum += digit;
        i *= 100;
    }

    if (sum % 10 == 0)
    {
        return true;
    }
    return false;
}
